### DISCUSSION TITLE
ESM vs CJS import woes

### DISCUSSION THREAD

**Discussion by larssn** [firebase/firebase-admin-node#2639] [Q&A] [answered]
So I'm trying to upgrade my modules to ESM, which I've done by setting `"type": "module"` in package.json, and `"module": "Node16"` in tsconfig.json's `compilerOptions`.

However I'm running into a few type issues.

If I import `Bucket` from @google-cloud/storage, it will correctly access the bucket in the ESM path.
But the `Bucket` type from firebase-admin's `storage().bucket()` will point to a CJS path.

Typescript will then complain that the two types are not compatible.

```ts
import { storage } from 'firebase-admin'
import { Bucket } from '@google-cloud/storage'
const bucket: Bucket = storage().bucket()
```
```
Type 'import("@google-cloud/storage/build/cjs/src/bucket").Bucket' is not assignable to type 'import("@google-cloud/storage/build/esm/src/bucket", { with: { "resolution-mode": "import" } }).Bucket'.
The types returned by 'storage.channel(...)' are incompatible between these types.
    Type 'import("@google-cloud/storage/build/cjs/src/channel").Channel' is not assignable to type 'import("@google-cloud/storage/build/esm/src/channel", { with: { "resolution-mode": "import" } }).Channel'.
      Types have separate declarations of a private property 'createMethod'.ts(2322)
```

How do I solve this "correctly"?

I can solve the above `Bucket` type problem by doing:
```ts
import type { Storage } from 'firebase-admin/storage'
type Bucket = ReturnType<Storage['bucket']>
```
but this quickly gets very complicated, as the next type I need to extract is `UploadOptions` which is 
the return type found on `Bucket.upload` which is a `Promise<UploadOptions>`. No idea how to extract that type, and I rather wouldn't.

Help would be appreciated.

**Comment by larssn** [comment, marked as answer]
So a workaround is doing:

```ts
import type { Bucket } from '@google-cloud/storage/build/cjs/src/bucket.d.ts'
```

And then adding the `bucket.d.ts` to `paths` in tsconfig's `compilerOptions:

```json
"paths": {
  "@google-cloud/storage/build/cjs/src/bucket.d.ts": ["./node_modules/@google-cloud/storage/build/cjs/src/bucket.d.ts"]
}
```

This basically screams "fix me later".

**Reply by bartekpacia** [reply]
I would love to see a better solution to this...

**Reply by risalfajar** [reply]
You should not mark a workaround as an answer

**Comment by risalfajar** [comment]
Please reopen this, this also happened to me when importing:
```ts
import type { File } from "@google-cloud/storage";
```

Error message:
```console
TS2322: Type
import("/WebstormProjects/firebase/functions/node_modules/@google-cloud/storage/build/cjs/src/file").File
is not assignable to type
import("/WebstormProjects/firebase/functions/node_modules/@google-cloud/storage/build/esm/src/file", { with: { "resolution-mode": "import" } }).File
Property #private in type File refers to a different member that cannot be accessed from within type File
```
